enhancements/local-storage/automatic-detection.md

```diff
@@ -74,125 +77,244 @@ The idea is to introduce a new Custom Resource, `AutoDetectVolumes`, that enable
+	// to contain at least one of these strings.
+	// +optional
+	Vendors []string `json:"vendors"`
 }
-```
 
-- Once the `AutoDetectVolume` is created, the controller will:
-  - configure the local-static-provisioner to make a new StorageClass based on certain directories on the selected nodes.
-  - assign diskmaker daemons to the selected nodes.
-- The diskmaker daemon will find devices that match the disovery policy and symlink them into the directory that the local-static-provisioner is watching.
+type AutoDetectVolumeStatus struct {
+	Status string `json:"status"`
+}
 
+type LocalVolumeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscovery `json:"items"`
+}
 
-### Test Plan
+type LocalVolumeDiscovery struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoverySpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryStatus `json:"status,omitempty"`
+}
 
-- The integration tests for the LSO already exist. These tests will need to be updated to test this feature.
-- The tests must ensure that detection of devices are working/updating correctly.
-- The tests must ensure that data corruption are not happening during auto detection of devices.
+type LocalVolumeDiscoverySpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
 
-### Graduation Criteria
+type LocalVolumeDiscoveryStatus struct {
+	Phase string `json:"phase,omitempty"`
+}
 
-- Documentation exists for the behaviour of each configuration item.
-- Unit and End to End tests coverage is sufficient.
+type NodeDeviceMap List struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []NodeDeviceMap `json:"items"`
+}
+
+type NodeDeviceMap struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              NodeDeviceMapSpec   `json:"spec"`
+	Status            NodeDeviceMapStatus `json:"status,omitempty"`
+}
+
+type NodeDeviceMapSpec struct {
+	NodeName string `json:"nodeName"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type NodeDeviceMapStatus struct {
+	Phase string `json:"phase,omitempty"`
+	// Blockdevices will be containing the output of following command unmarshalled into the structured
+	// "lsblk -o name,model,vendor,kname,pkname,type,uuid,tran,ro,rm,parttype,serial,size,rota,mountpoint"
+	Blockdevices []Device `json:"blockdevices"`
+}
+
+type Device struct {
```

Comment:
Another property we need is whether it has a filesystem. For example, in Rook this is detected with [udevadm](https://github.com/rook/rook/blob/b95e2a4c3efbcf55c2ef5db9a30abf9fc1a5cb5f/pkg/util/sys/device.go#L242). This could be used to determine if a file or block mode PV can be created on the device. Or do we only care about block-mode devices? If so, we would need to detect this, but then just wouldn't return devices with filesystems. For now I'd suggest we start with only block-mode, but keep the file-mode scenario in mind.

Reply:
yeah, I added the fstype in the output.

Reply:
Yeah, makes sense to include it in the discovery so users can reason about why devices aren't picked up by the `AutoDetectVolume`